### PR TITLE
Expose the in-memory data as SpanData.

### PR DESCRIPTION
**Pull request by carlosalberto**
This PR intends to be a follow up to #485 and to #440 

In addition to exposing the `Span` data as Proto Objects, we'd be exposing them as `SpanData` which, even though it was removed, proved to be a clean and clear way to expose `Span` in read-only fashion (and, if we happen to end up having the aforementioned `SpanData` class back in the SDK, we can simply 'forward' it here).

The advantage, IHMO, is that we won't have to expose many functions to help the user test:

```java
List<SpanData> spans = tracing.getFinishedSpanDataItems();
assertThat(spans.get(0).getContext().getTraceId()).isEqualTo(myTraceId);
assertThat(spans.get(0).getLinks().get(0).getAttributes()).hasSize(3);
```

(At the moment I'm using the default `Resource`, `Tracestate` and `TraceFlags` objects, but that will remain as a TODO to fix).

### CODE REVIEW

**Comment by jkwatson** [comment]
This would also be something that is super useful for non-protobuf based exporters. For example, the NewRelic exporter needs to generate JSON at the end of the day, which now requires a bunch of bending-over-backwards to extract the data from the protobuf representation (especially the ids). Would it make sense to have the SpanExporter interface just provide `SpanData` instances, rather than always converting to protobufs?

File: sdk_contrib/inmemory_export/src/main/java/io/opentelemetry/sdk/contrib/trace/export/SpanDataUtils.java [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+// TODO - Properly set TraceFlags/Tracestate for SpanContext and Links.
+final class SpanDataUtils {
+  private SpanDataUtils() {}
+
+  public static SpanData getFromProto(Span span) {
```
Comment (tylerbenson) [inline sdk_contrib/inmemory_export/src/main/java/io/opentelemetry/sdk/contrib/trace/export/SpanDataUtils.java]: why the suffix `FromProto` for these methods?  That implies some relationship to protobuf for me...  Perhaps `FromSpan` instead?

Comment (jkwatson) [reply sdk_contrib/inmemory_export/src/main/java/io/opentelemetry/sdk/contrib/trace/export/SpanDataUtils.java]: The Span instance being passed in is a protobuf instance. Maybe name the parameter `protobufSpan` to make it clear?

Comment (carlosalberto) [reply sdk_contrib/inmemory_export/src/main/java/io/opentelemetry/sdk/contrib/trace/export/SpanDataUtils.java]: Correct, it's a `io.opentelemetry.proto.trace.v1.Span` instance ;) Yeah, I could actually use `protoSpan` to make things more obvious :)

Comment (tylerbenson) [reply sdk_contrib/inmemory_export/src/main/java/io/opentelemetry/sdk/contrib/trace/export/SpanDataUtils.java]: Ah, I missed that part.

**Comment by bogdandrutu** [comment]
@jkwatson having the SpanExporter pipeline working with SpanData instead of proto is a valid option, which also removes a dependency on proto.

I am not necessary against that idea, but we thought it is simpler to use proto. Still I would not copy the SpanData as it was with a lot of checks and copies of the Collections/Maps, and probably would make the ctor package protected for the moment to allow us to not allocate any extra objects during conversion.

**Comment by jkwatson** [comment]
Could we just make the API Span have read methods on it? Then the exporters could just consume what the SDK is producing with no conversion needed at all.

**Comment by bogdandrutu** [comment]
I am strongly against this idea for multiple reasons that we mentioned previously. The major one being that that forces all the implementation to keep all the data in-memory until the Span ends which some of the vendors may want to avoid by implementing a streaming version.

We actually can add more readers on the ReadableSpan interface if we want.

**Comment by jkwatson** [comment]
As long as what is passed down to SpanExporters is generally usable by non-protobuf-based exporters, I'm fine with most anything. And, I agree that the less conversion-between-object formats that we can require, the better.  

Will Exporters ever be given non-complete spans, or will that only be sent to a separate streaming API?

**Comment by carlosalberto** [comment]
> We actually can add more readers on the ReadableSpan interface if we want.

I like this one, and I think it would work just fine.

**Comment by carlosalberto** [comment]
> Will Exporters ever be given non-complete spans, or will that only be sent to a separate streaming API?

The current SDK implementation is not a streaming one, so we will only expose complete `Span`s   (again, this is only from the current SDK side ;) )

**Comment by jkwatson** [comment]
So, is the idea that instead of proto Spans, we give the exporters ReadableSpan instances, with enough getters on the interface to be able to do exports? That sounds great to me!

**Comment by bogdandrutu** [comment]
I keep think about this. Don't know if exposing to the exporter ReadableSpan or SpanData if we want to not use proto there. The problem with ReadableSpan is that it holds to all the memory referenced by any lazy event/attribute/link recorded.

So the trade-off is:
1. If we do use ReadableSpan then we may reference objects from the user for some long time.
2. If we do use SpanData (or proto) we pay the cost of constructing these objects. The cost of formatting the lazy events we pay anyway at one point.

**Comment by carlosalberto** [comment]
> If we do use ReadableSpan then we may reference objects from the user for some long time.

There's also the thing about an exporter potentially being able to modify the `Span`, compared to the current approach, where the user gets at this a point a read-only version.

**Comment by bogdandrutu** [comment]
@jkwatson this will become a no-op after your change to use SpanData instead of the protos.

**Comment by jkwatson** [comment]
Agreed! I think changing to exporters using SpanData will simplify several things. We will make it so!

**Comment by jkwatson** [comment]
yeah, I think this should be closed. It will come for "free" when we get our big SpanData exporter update landed.

**Pull request closed without merging**